scripts/move_ean8_products_to_new_path.pl

```diff
@@ -0,0 +1,347 @@
+use ProductOpener::Images qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Mail qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::Food qw/:all/;
+use ProductOpener::Ingredients qw/:all/;
+use ProductOpener::Images qw/:all/;
+use ProductOpener::Data qw/:all/;
+
+use CGI qw/:cgi :form escapeHTML/;
+use URI::Escape::XS;
+use Storable qw/dclone/;
+use Encode;
+use JSON::PP;
+
+use Data::Dumper;
+
+sub normalize_code_zeroes($code) {
+
+	# Remove leading zeroes
+	$code =~ s/^0+//;
+
+	# Add leading zeroes to have at least 13 digits
+	while (length($code) < 13) {
+		$code = "0" . $code;
+	}
+
+	# Remove leading zeroes for EAN8s
+	if ((length($code) eq 13) and ($code =~ /^00000/)) {
+		$code = $';
+	}
+
+	return $code;
+}
+
+# This script includes a new_split_code() function so that we can run it even if Products.pm split_code() is old
+# This is useful in particular for preparations before the move, e.g. to see which products would be affected
+
+sub is_valid_code ($code) {
+	# Return an empty string if $code is undef
+	return '' if !defined $code;
+	return $code =~ /^\d{4,24}$/;
+}
+
+sub new_split_code ($code) {
+
+	# Require at least 4 digits (some stores use very short internal barcodes, they are likely to be conflicting)
+	if (not is_valid_code($code)) {
+		return "invalid";
+	}
+
+	# Pad code with 0s if it has less than 13 digits
+	while (length($code) < 13) {
+		$code = "0" . $code;
+	}
+
+	# First splits into 3 sections of 3 numbers and the last section with the remaining numbers
+	my $path = $code;
+	if ($code =~ /^(.{3})(.{3})(.{3})(.*)$/) {
+		$path = "$1/$2/$3/$4";
+	}
+	return $path;
+}
+
+sub new_product_path_from_id ($product_id) {
+
+	my $product_id_without_server = $product_id;
+	$product_id_without_server =~ s/(.*)://;
+
+	if (    (defined $server_options{private_products})
+		and ($server_options{private_products})
+		and ($product_id_without_server =~ /\//))
+	{
+		return $` . "/" . new_split_code($');
+	}
+	else {
+		return new_split_code($product_id_without_server);
+	}
+
+}
+
+sub ensure_dir_created_or_die ($path, $mode = oct(755)) {
+	# search base directory
+	my $prefix;
+	my $suffix;
+	my @base_dirs = ($data_root, $www_root);
+	foreach my $prefix_candidate (@base_dirs) {
+		if ($path =~ /^$prefix_candidate/) {
+			$prefix = $prefix_candidate;
+			$suffix = $';
+			last;
+		}
+	}
+	if (!defined $prefix) {
+		die("Could not create $path, no corresponding base directory found in " . join(":", @base_dirs));
+		return;
+	}
+	# ensure the rest of the path
+	foreach my $component (split(/\//, $suffix)) {
+		$prefix .= "/$component";
+		(-e $prefix) or mkdir($prefix);
+	}
+	return (-e $path);
+}
+
+# Get a list of all products
+
+use Getopt::Long;
+
+my @products = ();
+my $move = 0;
+
+GetOptions('products=s' => \@products, 'move' => \$move);
+@products = split(/,/, join(',', @products));
+
+my $d = 0;
+
+open(my $log, ">>", "$data_root/logs/move_ean8_products_to_new_path.log");
```

Comment:
nit: you should normally be using the Path.pm module !

Reply:
I need the script to work on the old code (so that I can run it in dry-run mode on OBF, OPF, OPFF, to see which products are affected), that's why it does not use some of the newer things like Path.pm